hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/DatanodeIdYaml.java

```diff
@@ -82,6 +83,12 @@ public static DatanodeDetails readDatanodeIdFile(File path)
           .setIpAddress(datanodeDetailsYaml.getIpAddress())
           .setHostName(datanodeDetailsYaml.getHostName())
           .setCertSerialId(datanodeDetailsYaml.getCertSerialId());
+      if (datanodeDetailsYaml.getPersistedOpState() != null) {
+        builder.setPersistedOpState(HddsProtos.NodeOperationalState.valueOf(
+            datanodeDetailsYaml.getPersistedOpState()));
+      }
+      builder.setPersistedOpStateExpiry(
```

Comment:
Shouldn't this be also inside the if != null ? , I am going to presume that even if we have no data, datanodeDetailsYaml.getPersistedOpStateExpiryEpochSec is going to return zero, but since the expiry is a long, JVM would initialize it to zero by default, would it not?. So curious why this check is outside the if.

Reply:
If I recall (I wrote this code a month or so ago and forgot to create the PR) the reason the OpState is wrapped in a IF block is because I need take the string read from the YAML file and turn it into the enum. The enum didn't like me calling valueOf(null), so I only make that call if its non-null. All the other parameters in the builder are strings except for the two new ones (enum and long), so they never really had to handle the null case before.

You are correct, in that if there is no data, the "getPersistedOpStateExpiryEpochSec" will end up as zero. That is OK, as SCM treats zero as "no expiry" right now.

I've got similar logic at the end of that class too to handle the turning the enum into a string:

```
    String persistedOpString = null;
    if (datanodeDetails.getPersistedOpState() != null) {
      persistedOpString = datanodeDetails.getPersistedOpState().name();
    }
    return new DatanodeDetailsYaml(
        datanodeDetails.getUuid().toString(),
        datanodeDetails.getIpAddress(),
        datanodeDetails.getHostName(),
        datanodeDetails.getCertSerialId(),
        persistedOpString,
        datanodeDetails.getPersistedOpStateExpiryEpochSec(),
        portDetails);
  }
```

I think the shorter summary is that the enums needs special handling around the nulls while the long and Strings behave OK without wrapping them in IF blocks.